askama_shared/src/generator.rs

```diff
@@ -60,14 +61,135 @@ fn build_template(ast: &syn::DeriveInput) -> Result<String, CompileError> {
+        config.suppress_whitespace,
+    )
+    .build(&contexts[input.path.as_path()])?;
     if input.print == Print::Code || input.print == Print::All {
         eprintln!("{}", code);
     }
     Ok(code)
 }
 
+#[derive(Default)]
+pub(crate) struct TemplateArgs {
+    pub(crate) source: Option<Source>,
+    pub(crate) print: Print,
+    pub(crate) escaping: Option<String>,
+    pub(crate) ext: Option<String>,
+    pub(crate) syntax: Option<String>,
+    pub(crate) config_path: Option<String>,
+}
+
+impl TemplateArgs {
+    fn new(ast: &'_ syn::DeriveInput) -> Result<Self, CompileError> {
+        // Check that an attribute called `template()` exists once and that it is
+        // the proper type (list).
+        let mut template_args = None;
+        for attr in &ast.attrs {
+            let ident = match attr.path.get_ident() {
+                Some(ident) => ident,
+                None => continue,
+            };
+
+            if ident == "template" {
+                if template_args.is_some() {
+                    return Err("duplicated 'template' attribute".into());
+                }
+
+                match attr.parse_meta() {
+                    Ok(syn::Meta::List(syn::MetaList { nested, .. })) => {
+                        template_args = Some(nested);
+                    }
+                    Ok(_) => return Err("'template' attribute must be a list".into()),
+                    Err(e) => return Err(format!("unable to parse attribute: {}", e).into()),
+                }
+            }
+        }
+        let template_args =
+            template_args.ok_or_else(|| CompileError::from("no attribute 'template' found"))?;
+
+        let mut args = Self::default();
+        // Loop over the meta attributes and find everything that we
+        // understand. Return a CompileError if something is not right.
+        // `source` contains an enum that can represent `path` or `source`.
+        for item in template_args {
+            let pair = match item {
+                syn::NestedMeta::Meta(syn::Meta::NameValue(ref pair)) => pair,
+                _ => {
+                    return Err(format!(
+                        "unsupported attribute argument {:?}",
+                        item.to_token_stream()
+                    )
+                    .into())
+                }
+            };
+            let ident = match pair.path.get_ident() {
+                Some(ident) => ident,
+                None => unreachable!("not possible in syn::Meta::NameValue(…)"),
+            };
+
+            if ident == "path" {
+                if let syn::Lit::Str(ref s) = pair.lit {
+                    if args.source.is_some() {
+                        return Err("must specify 'source' or 'path', not both".into());
+                    }
+                    args.source = Some(Source::Path(s.value()));
+                } else {
+                    return Err("template path must be string literal".into());
+                }
```

Comment:
I know it's just old code being moved. But the nested ifs could be replaced with a single match. _(Just did a quick copy, paste, format by hand.)_

@djc which do you prefer?

```suggestion
                match pair.lit {
                    syn::Lit::Str(ref s) = pair.lit if args.source.is_some() => return Err("must specify 'source' or 'path', not both".into()),
                    syn::Lit::Str(ref s) = pair.lit => args.source = Some(Source::Path(s.value())),
                    _ => return Err("template path must be string literal".into()),
                }
```

Reply:
Let's do it in a separate PR. I think a `match` expression makes sense here, I think the `args.source` case could maybe be handled with an `ok_or()` or `ok_or_else()`?

Reply:
Separate PR it is 👍 Not sure how it would work with `ok_or()`/`ok_or_else()` as the error case is if it's `Some` not `None`? :)

Reply:
Ah yes, that would be the other way around.

Reply:
Depending on the formatting rustfmt comes up with, an inner `match` might be nicer than using guards on the outer match arms.